### PR TITLE
Remove applicant id from application programs view URL

### PR DESCRIPTION
**Pull request by yotommy**
### Description

Remove applicant id from application programs view URL.

This is done in the usual way: 
* define a new `ApplicantRoutes.view()` method that delegates to the existing method, 
* which is renamed to `viewWithApplicantId()`, and
* revise the callsites to use the new applicant route method.

This PR has a complication. Before this PR, the existing route was:

```
/applicants/:applicantId/programs/:programId
```

Since Play does not permit overloaded controller methods, when we remove the leading `/applicants/:applicantId` path components, that results in a conflict with this route:

```
/programs/:programSlug
```

Therefore the new `ApplicantProgramsController.view()` method needs to disambiguate between the two cases. At this point, we no longer need a separate `DeepLinkController`, so this class has been transformed into a `ProgramSlugHandler` that is created by injection.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers).
- [x] Created unit and/or browser tests which fail without the change (if possible)

### Issue(s) this completes

Relates to #5576

### CODE REVIEW

**Comment by yotommy** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6063** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6063?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #5946** <a href="https://app.graphite.dev/github/pr/civiform/civiform/5946?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #6115** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6115?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #6060** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6060?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #5979** <a href="https://app.graphite.dev/github/pr/civiform/civiform/5979?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #5905** <a href="https://app.graphite.dev/github/pr/civiform/civiform/5905?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/civiform/civiform/6115?utm_source=stack-comment).

**Comment by yotommy** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6063** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6063?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #5946** <a href="https://app.graphite.dev/github/pr/civiform/civiform/5946?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #6115** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6115?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #6060** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6060?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #5979** <a href="https://app.graphite.dev/github/pr/civiform/civiform/5979?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #5905** <a href="https://app.graphite.dev/github/pr/civiform/civiform/5905?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/civiform/civiform/6115?utm_source=stack-comment).

File: server/app/controllers/applicant/ApplicantProgramsController.java [hunk 1]
```diff
@@ -113,7 +118,8 @@ public CompletionStage<Result> index(Request request) {
   }
 
   @Secure
-  public CompletionStage<Result> view(Request request, long applicantId, long programId) {
+  public CompletionStage<Result> viewWithApplicantId(
+      Request request, long applicantId, long programId) {
```
Comment (bion) [inline server/app/controllers/applicant/ApplicantProgramsController.java:122]: **nitpick:** let's take this opportunity to rename this to `show` to fit the naming schema: https://github.com/civiform/civiform/wiki/Development-standards#html-routing-convention

File: server/app/controllers/applicant/ApplicantProgramsController.java [hunk 2]
```diff
@@ -162,6 +168,26 @@ public CompletionStage<Result> view(Request request, long applicantId, long prog
...
+  //
+  // Because the second use is public, this controller is not annotated as @Secure. For the first
+  // use, the delegated-to method *is* annotated as such.
+  public CompletionStage<Result> view(Request request, String programParam) {
+    if (StringUtils.isNumeric(programParam)) {
+      // The path parameter specifies a program by (numeric) id.
+
+      // The route for this action should only be computed if the applicant ID is available in the
+      // session.
+      long applicantId = getApplicantId(request).orElseThrow();
```
Comment (bion) [inline server/app/controllers/applicant/ApplicantProgramsController.java:184]: **issue (blocking):** throwing here will produce a 500 (server error), indicating a problem in the server code. If this throws though it's the result of a bad request -- an HTTP client without a session attempting to access a route they can't be authenticated for.

One option would be to return a 4XX level status code instead, which will at least help our monitoring be more accurate.

Something else to consider is when this may occur -- e.g. if someone bookmarks the URL for an application they're working on and then their session expires. Is there a way to redirect them to the identity provider instead?

File: server/app/controllers/applicant/ApplicantRoutes.java [hunk 3]
```diff
@@ -49,4 +49,24 @@ public Call index(CiviFormProfile profile, long applicantId) {
...
+
+  /**
+   * Returns the route corresponding to the applicant view action.
+   *
+   * @param profile - Profile corresponding to the logged-in user (applicant or TI).
+   * @param applicantId - ID of applicant for whom the action should be performed.
+   * @param programId - ID of program to view
+   * @return Route for the program view action
+   */
+  public Call view(CiviFormProfile profile, long applicantId, long programId) {
```
Comment (bion) [inline server/app/controllers/applicant/ApplicantRoutes.java:61]: **nitpick:** s/view/show to follow route naming convention

File: server/app/modules/MainModule.java [hunk 4]
```diff
@@ -84,4 +90,15 @@ public OidcClientProviderParams provideOidcClientProviderParams(
...
+  @Provides
+  public ProgramSlugHandler provideProgramSlugHandler(
+      HttpExecutionContext httpContext,
+      ApplicantService applicantService,
+      ProfileUtils profileUtils,
+      ProgramService programService,
+      LanguageUtils languageUtils) {
+    return new ProgramSlugHandler(
+        httpContext, applicantService, profileUtils, programService, languageUtils);
+  }
```
Comment (bion) [inline server/app/modules/MainModule.java:103]: **question (blocking):** most injected classes don't need a provider. Why is this needed? Is there an injection error without it?

**Review by bion — CHANGES_REQUESTED** [review]

File: server/app/modules/MainModule.java (hunk 4 above)
Comment (yotommy) [reply server/app/modules/MainModule.java:103]: TIL, thanks. Removed.

File: server/app/controllers/applicant/ApplicantProgramsController.java (hunk 1 above)
Comment (yotommy) [reply server/app/controllers/applicant/ApplicantProgramsController.java:122]: Done

File: server/app/controllers/applicant/ApplicantRoutes.java (hunk 3 above)
Comment (yotommy) [reply server/app/controllers/applicant/ApplicantRoutes.java:61]: Done

File: server/app/controllers/applicant/ApplicantProgramsController.java (hunk 2 above)
Comment (yotommy) [reply server/app/controllers/applicant/ApplicantProgramsController.java:184]: This last point is an excellent one that had not occurred to me. I've made the change here and will need to go back and make the corresponding change in #5946 before merging.

File: server/app/controllers/CiviFormController.java [hunk 5]
```diff
@@ -72,14 +74,27 @@ protected CompletionStage<Void> checkProgramAuthorization(Http.Request request,
...
   protected Optional<Long> getApplicantId(Http.Request request) {
-    CiviFormProfileData profileData =
-        profileUtils.currentUserProfile(request).orElseThrow().getProfileData();
+    Optional<CiviFormProfile> profile = profileUtils.currentUserProfile(request);
+    if (profile.isEmpty()) {
+      return Optional.empty();
+    }
 
+    CiviFormProfileData profileData = profile.orElseThrow().getProfileData();
     if (profileData == null) {
```
Comment (bion) [inline server/app/controllers/CiviFormController.java:83]: **suggestion (non-blocking, if-minor):** i think could do?
```suggestion
    if (profile.map(CiviFormProfile::getProfileData).flatMap(Optional.ofNullable).isEmpty()) {
```

Comment (yotommy) [reply server/app/controllers/CiviFormController.java:83]: Nice, thanks
